### PR TITLE
qt replay

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

File: selfdrive/ui/qt/api.cc [hunk 1]
```diff
@@ -118,15 +122,15 @@ void RequestRepeater::sendRequest(QString requestURL){
   connect(reply, SIGNAL(finished()), this, SLOT(requestFinished()));
 }
 
-void RequestRepeater::requestTimeout(){
+void HttpRequest::requestTimeout(){
   reply->abort();
 }
 
-// This function should always emit something
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/api.cc:125]: restore the comment

File: selfdrive/ui/qt/api.hpp [hunk 2]
```diff
@@ -28,18 +28,19 @@ class CommaApi : public QObject {
 /**
  * Makes repeated requests to the request endpoint.
  */
-class RequestRepeater : public QObject {
+
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/api.hpp:31]: update the comment above

File: selfdrive/ui/qt/api.hpp [hunk 3]
```diff
@@ -28,18 +28,19 @@ class CommaApi : public QObject {
...
+class HttpRequest : public QObject {
   Q_OBJECT
 
 public:
-  explicit RequestRepeater(QWidget* parent, QString requestURL, int period = 10, const QString &cache_key = "", bool disableWithScreen = true);
+  explicit HttpRequest(QWidget* parent, QString requestURL, const QString &cache_key = "", int period = 0, bool disableWithScreen = true);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/api.hpp:36]: did anything rely on that 10s default?

File: selfdrive/ui/replay/.gitignore [hunk 4]
```diff
@@ -0,0 +1,4 @@
+.*.swp
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/.gitignore:1]: remove this file and add replay to the gitignore in selfdrive/ui

File: selfdrive/ui/SConscript [hunk 5]
```diff
@@ -28,8 +28,21 @@ qt_env.Program("qt/spinner", ["qt/spinner.cc"], LIBS=base_libs)
 # build main UI
 qt_src = ["main.cc", "ui.cc", "paint.cc", "sidebar.cc", "#phonelibs/nanovg/nanovg.c",
           "qt/window.cc", "qt/home.cc", "qt/api.cc", "qt/offroad/settings.cc",
-          "qt/offroad/onboarding.cc"]
-qt_env.Program("_ui", qt_src, LIBS=qt_libs)
+          "qt/offroad/onboarding.cc", "replay/Unlogger.cpp", "../../tools/clib/FrameReader.cpp",
```
Comment (adeebshihadeh) [inline selfdrive/ui/SConscript:31]: build a separate replay lib

File: selfdrive/ui/SConscript [hunk 6]
```diff
@@ -28,8 +28,21 @@ qt_env.Program("qt/spinner", ["qt/spinner.cc"], LIBS=base_libs)
...
+replay_libs = ['avutil', 'avcodec', 'avformat', 'swscale', 'bz2']
+
+qt_env.Program("_ui", qt_src, LIBS=qt_libs + replay_libs)
+
+# build headless replay
+replay_src = ["replay/main.cc", "replay/replay.cc", "replay/Unlogger.cpp",
+              "../../tools/clib/FrameReader.cpp", "replay/FileReader.cpp",
+              "qt/api.cc", "#phonelibs/nanovg/nanovg.c"]
+
+qt_env.Program("replay/replay", replay_src, LIBS=qt_libs + replay_libs)
```
Comment (adeebshihadeh) [inline selfdrive/ui/SConscript:45]: only build this on PC

File: selfdrive/ui/qt/api.cc [hunk 7]
```diff
@@ -72,29 +72,33 @@ QString CommaApi::create_jwt(QVector<QPair<QString, QJsonValue>> payloads, int e
...
-
   networkTimer = new QTimer(this);
   networkTimer->setSingleShot(true);
   networkTimer->setInterval(20000);
   connect(networkTimer, SIGNAL(timeout()), this, SLOT(requestTimeout()));
 
+  if(period_seconds <= 0){
+    sendRequest(requestURL);
+  }
+  else{
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/api.cc:88]: same line

File: selfdrive/ui/SConscript [hunk 8]
```diff
@@ -28,8 +28,22 @@ qt_env.Program("qt/spinner", ["qt/spinner.cc"], LIBS=base_libs)
...
+          "qt/offroad/onboarding.cc", "replay/Unlogger.cpp", "../../tools/clib/FrameReader.cpp",
+          "replay/FileReader.cpp", "replay/replay.cc"]
+
+qt_env['CPPPATH'] += ["#tools/clib"]
+qt_env['CXXFLAGS'] += ["-Wno-deprecated-declarations"]
+replay_libs = ['avutil', 'avcodec', 'avformat', 'swscale', 'bz2']
+
+qt_env.Program("_ui", qt_src, LIBS=qt_libs + replay_libs)
+
+if arch == 'aarch64':
```
Comment (adeebshihadeh) [inline selfdrive/ui/SConscript:40]: wrong aarch

File: selfdrive/ui/main.cc [hunk 9]
```diff
@@ -3,6 +3,12 @@
 #include "qt/window.hpp"
 #include "qt/qt_window.hpp"
 
+#include <QThread>
```
Comment (adeebshihadeh) [inline selfdrive/ui/main.cc:6]: revert changes to this file

File: selfdrive/ui/qt/home.cc [hunk 10]
```diff
@@ -299,6 +299,7 @@ void GLWindow::timerUpdate() {
   sound.volume = util::map_val(ui_state.scene.car_state.getVEgo(), 0.f, 20.f,
                                Hardware::MIN_VOLUME, Hardware::MAX_VOLUME);
 
+
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/home.cc:302]: revert changes to this file

File: selfdrive/ui/replay/Unlogger.cpp [hunk 11]
```diff
@@ -165,21 +153,41 @@ void Unlogger::process() {
...
+          msg.setRoot(e);
+          auto words = capnp::messageToFlatArray(msg);
+          auto bytes = words.asBytes();
 
-        // TODO: Can PubSocket take a const char?
-        (*it)->send((char*)bytes.begin(), bytes.size());
+          // TODO: Can PubSocket take a const char?
+          (*it)->send((char*)bytes.begin(), bytes.size());
+        }
+        else{
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/Unlogger.cpp:175]: same line

File: selfdrive/ui/replay/replay.cc [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  camera_paths = doc["cameras"].toArray();
+  log_paths = doc["logs"].toArray();
+
+  // add first segment
+  addSegment(0);
+}
+
+bool Replay::addSegment(int i){
+  if (lrs.find(i) == lrs.end()) {
+    QString fn = QString("http://data.comma.life/%1/%2/rlog.bz2").arg(route).arg(i);
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.cc:36]: remove this

File: selfdrive/ui/replay/replay.hpp [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <QFile>
+#include <QQueue>
+#include <QJsonArray>
+#include <QJsonObject>
+#include <QJsonDocument>
+
+#include <capnp/dynamic.h>
+
+#include "../qt/api.hpp"
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.hpp:11]:
```suggestion
#include "qt/api.hpp"
```

File: selfdrive/ui/replay/replay.cc [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+  http = new HttpRequest(this, "https://api.commadotai.com/v1/route/" + route.replace("/", "|") + "/files", "Route_HttpRequest", 0, false);
+  QObject::connect(http, SIGNAL(receivedResponse(QString)), this, SLOT(parseResponse(QString)));
+}
+
+void Replay::parseResponse(QString response){
+  response = response.trimmed();
+  QJsonDocument doc = QJsonDocument::fromJson(response.toUtf8());
+
+  if (doc.isNull()) {
+    qDebug() << "JSON Parse failed on getting past drives statistics";
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.cc:16]: drive stats?

File: selfdrive/ui/replay/main.cc [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  QApplication a(argc, argv);
+
+  QString route(argv[1]);
+  route = route.replace("|", "/");
+  if (route == "") {
+    printf("Usage: ./replay \"route\"\n");
+    return 1;
+  }
+
+  int seek = QString(argv[2]).toInt();
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/main.cc:15]: what happens when there's only one arg?

File: selfdrive/ui/qt/api.cc [hunk 16]
```diff
@@ -72,29 +72,32 @@ QString CommaApi::create_jwt(QVector<QPair<QString, QJsonValue>> payloads, int e
...
-  timer->start(period_seconds * 1000);
-
   networkTimer = new QTimer(this);
   networkTimer->setSingleShot(true);
   networkTimer->setInterval(20000);
   connect(networkTimer, SIGNAL(timeout()), this, SLOT(requestTimeout()));
 
+  if(period_seconds <= 0){
+    sendRequest(requestURL);
+  } else{
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/api.cc:87]: whitespace

File: selfdrive/ui/SConscript [hunk 17]
```diff
@@ -14,7 +14,8 @@ if arch == "Darwin":
 widgets_src = ["qt/widgets/input.cc", "qt/widgets/drive_stats.cc", "qt/sound.cc",
                "qt/widgets/ssh_keys.cc", "qt/widgets/toggle.cc", "qt/widgets/controls.cc",
                "qt/widgets/offroad_alerts.cc", "qt/widgets/setup.cc", "qt/widgets/keyboard.cc",
-               "qt/widgets/scrollview.cc", "#phonelibs/qrcode/QrCode.cc"]
+               "qt/widgets/scrollview.cc", "#phonelibs/qrcode/QrCode.cc", "qt/api.cc",
+               "#phonelibs/nanovg/nanovg.c"]
```
Comment (adeebshihadeh) [inline selfdrive/ui/SConscript:18]: move nanovg back to qt_src

File: selfdrive/ui/qt/api.hpp [hunk 18]
```diff
@@ -26,22 +26,21 @@ class CommaApi : public QObject {
 };
 
 /**
- * Makes repeated requests to the request endpoint.
+ * Makes request to the request endpoint.
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/api.hpp:29]: grammar

File: selfdrive/ui/qt/request_repeater.hpp [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+#include "home.hpp"
+
+class RequestRepeater : public HttpRequest {
+
+public:
+  RequestRepeater(QWidget *parent, QString requestURL, const QString &cache_key = "", int period_seconds = 0, bool disableWithScreen = true);
+  bool disableWithScreen;
+};
+
+
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/request_repeater.hpp:11]: remove these trailing new lines

File: selfdrive/ui/SConscript [hunk 20]
```diff
@@ -55,3 +57,17 @@ if arch != 'aarch64' and "BUILD_SETUP" in os.environ:
...
     obj = qt_env.Object(f"qt/setup/installer_{name}.o", ["qt/setup/installer.cc"], CPPDEFINES=d)
     qt_env.Program(f"qt/setup/installer_{name}", obj, LIBS=qt_libs, CPPDEFINES=d)
+
+# build headless replay
+if arch == 'x86_64':
+  qt_env['CPPPATH'] += ["#tools/clib"]
+  qt_env['CXXFLAGS'] += ["-Wno-deprecated-declarations"]
+
+  replay_lib_src = ["replay/replay.cc", "replay/Unlogger.cpp",
+                "#tools/clib/FrameReader.cpp", "replay/FileReader.cpp"]
```
Comment (adeebshihadeh) [inline selfdrive/ui/SConscript:67]: indentation

File: selfdrive/ui/replay/main.cc (hunk 15 above)
Comment (adeebshihadeh) [reply selfdrive/ui/replay/main.cc:15]: let's remove this and just do the same thing as unlogger.py in the next PR

File: selfdrive/ui/replay/Unlogger.cpp [hunk 21]
```diff
@@ -165,21 +153,40 @@ void Unlogger::process() {
...
-
-        // TODO: Can PubSocket take a const char?
-        (*it)->send((char*)bytes.begin(), bytes.size());
+        if (sm == nullptr){
+          capnp::MallocMessageBuilder msg;
+          msg.setRoot(e);
+          auto words = capnp::messageToFlatArray(msg);
+          auto bytes = words.asBytes();
+
+          // TODO: Can PubSocket take a const char?
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/Unlogger.cpp:172]: resolve the TODO

File: selfdrive/ui/replay/Unlogger.cpp [hunk 22]
```diff
@@ -165,21 +153,40 @@ void Unlogger::process() {
...
+          sm->update_msgs(nanos_since_boot(), messages);
+        }
       }
       ++eit;
+
+      float time_to_end = ((events->lastKey() - eit.key())/1e9);
+
+      if (time_to_end < 10.0){
+        emit loadSegment();
+        blockSignals(true);
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/Unlogger.cpp:186]: why blockSignals? this seems pretty messy

File: selfdrive/ui/SConscript [hunk 23]
```diff
@@ -55,3 +57,17 @@ if arch != 'aarch64' and "BUILD_SETUP" in os.environ:
...
     obj = qt_env.Object(f"qt/setup/installer_{name}.o", ["qt/setup/installer.cc"], CPPDEFINES=d)
     qt_env.Program(f"qt/setup/installer_{name}", obj, LIBS=qt_libs, CPPDEFINES=d)
+
+# build headless replay
+if arch == 'x86_64':
+  qt_env['CPPPATH'] += ["#tools/clib"]
+  qt_env['CXXFLAGS'] += ["-Wno-deprecated-declarations"]
+
+  replay_lib_src = ["replay/replay.cc", "replay/Unlogger.cpp",
+                    "#tools/clib/FrameReader.cpp", "replay/FileReader.cpp"]
```
Comment (adeebshihadeh) [inline selfdrive/ui/SConscript:67]: sort this by directory

File: selfdrive/ui/SConscript [hunk 24]
```diff
@@ -55,3 +57,17 @@ if arch != 'aarch64' and "BUILD_SETUP" in os.environ:
...
+  replay_libs = ['avutil', 'avcodec', 'avformat', 'swscale', 'bz2']
+  replay = qt_env.Library("qt_replay", replay_lib_src, LIBS=base_libs)
+  replay = [replay] + replay_libs + qt_libs
+
+  replay_src = ["replay/main.cc"]
+  qt_env.Program("replay/replay", replay_src, LIBS=qt_libs + replay)
```
Comment (adeebshihadeh) [inline selfdrive/ui/SConscript:73]:
```suggestion
  replay_lib = qt_env.Library("qt_replay", replay_lib_src, LIBS=base_libs)
  replay_libs = qt_libs + ['avutil', 'avcodec', 'avformat', 'swscale', 'bz2', replay_lib]
  qt_env.Program("replay/replay", ["replay/main.cc"], LIBS=replay_libs)
```

File: selfdrive/ui/qt/request_repeater.cc [hunk 25]
```diff
@@ -0,0 +1,14 @@
+#include "request_repeater.hpp"
+
+RequestRepeater::RequestRepeater(QWidget *parent, QString requestURL, const QString &cache_key, int period_seconds, bool disableWithScreen) :
+  HttpRequest(parent, requestURL, cache_key), disableWithScreen(disableWithScreen) {
+  QTimer* timer = new QTimer(this);
+  QObject::connect(timer, &QTimer::timeout, [=](){
+    if (GLWindow::ui_state.scene.started || reply != NULL ||
+      (!GLWindow::ui_state.awake && disableWithScreen)) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/request_repeater.cc:8]: indentation

File: selfdrive/ui/qt/request_repeater.cc [hunk 26]
```diff
@@ -0,0 +1,14 @@
...
+      return;
+    }
+    sendRequest(requestURL);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/request_repeater.cc:11]:
```suggestion
    if (!GLWindow::ui_state.scene.started && reply == NULL && (GLWindow::ui_state.awake || !disableWithScreen)) {
      sendRequest(requestURL);
    }
```

File: selfdrive/ui/replay/Unlogger.cpp [hunk 27]
```diff
@@ -57,23 +48,20 @@ Unlogger::Unlogger(Events *events_, QReadWriteLock* events_lock_, QMap<int, Fram
...
+  cl_device_id device_id = cl_get_device_id(CL_DEVICE_TYPE_DEFAULT);
+  cl_context context = CL_CHECK_ERR(clCreateContext(NULL, 1, &device_id, NULL, NULL, &err));
 
-        //qDebug() << "here" << tname.c_str() << type << cereal::Event::CONTROLS_STATE;
-        socks.insert(type, sock);
-      }
-    }
-  }
+  vipc_server = new VisionIpcServer("camerad", device_id, context);
+  vipc_server->create_buffers(VisionStreamType::VISION_STREAM_RGB_BACK, 4, true, 1164, 874);
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/Unlogger.cpp:58]: get the width and height from the image

File: selfdrive/ui/replay/Unlogger.hpp [hunk 28]
```diff
@@ -15,19 +17,23 @@ Q_OBJECT
...
   signals:
     void elapsed();
     void finished();
+    void loadSegment();
   private:
     Events *events;
     QReadWriteLock *events_lock;
     QMap<int, FrameReader*> *frs;
-    QMap<int, PubSocket*> socks;
+    QMap<std::string, PubSocket*> socks;
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/Unlogger.hpp:31]: use a PubMaster

File: selfdrive/ui/replay/Unlogger.cpp [hunk 29]
```diff
@@ -165,21 +153,40 @@ void Unlogger::process() {
...
+          messages.push_back({type, e});
+          sm->update_msgs(nanos_since_boot(), messages);
+        }
       }
       ++eit;
+
+      float time_to_end = ((events->lastKey() - eit.key())/1e9);
+
+      if (time_to_end < 10.0){
+        emit loadSegment();
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/Unlogger.cpp:185]: maybe have another thread running that just maintains a 3 segment window

File: selfdrive/ui/replay/Unlogger.cpp [hunk 30]
```diff
@@ -147,14 +141,8 @@ void Unlogger::process() {
...
-
-        auto ee = msg.getRoot<cereal::Event>();
-        ee.setLogMonoTime(nanos_since_boot());
-
-        if (e.which() == cereal::Event::ROAD_CAMERA_STATE) {
-          auto fr = msg.getRoot<cereal::Event>().getRoadCameraState();
+        if (type == "roadCameraState") {
+          auto fr = e.getRoadCameraState();
 
           // TODO: better way?
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/Unlogger.cpp:147]: resolve the TODO

File: selfdrive/ui/replay/Unlogger.cpp [hunk 31]
```diff
@@ -127,8 +115,14 @@ void Unlogger::process() {
         last_elapsed = tc;
       }
 
-      auto e = *eit;
-      auto type = e.which();
+      cereal::Event::Reader e = *eit;
+
+      capnp::DynamicStruct::Reader e_ds = static_cast<capnp::DynamicStruct::Reader>(e);
+      std::string type;
+      KJ_IF_MAYBE(e_, e_ds.which()){
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/Unlogger.cpp:122]: is this necessary?

File: selfdrive/ui/replay/replay.cc [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+
+void Replay::stream(int seek, SubMaster *sm){
+  QThread* thread = new QThread;
+  unlogger->moveToThread(thread);
+  QObject::connect(thread, &QThread::started, [=](){
+    unlogger->process(sm);
+  });
+  thread->start();
+
+  QObject::connect(unlogger, &Unlogger::loadSegment, [=](){
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.cc:66]: why not connect this in the constructor?

File: selfdrive/ui/replay/Unlogger.cpp (hunk 31 above)
Comment (iejMac) [reply selfdrive/ui/replay/Unlogger.cpp:122]: i think this is the default way to remove the Maybe prefix

File: selfdrive/ui/replay/replay.cc (hunk 32 above)
Comment (iejMac) [reply selfdrive/ui/replay/replay.cc:66]: easier to call addSegment and iterate seg_add also no access to addSegment function in unlogger

File: selfdrive/ui/replay/replay.hpp [hunk 33]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <capnp/dynamic.h>
+
+#include "qt/api.hpp"
+#include "Unlogger.hpp"
+#include "FileReader.hpp"
+#include "FrameReader.hpp"
+#include "visionipc_server.h"
+
+class Replay : public QWidget {
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.hpp:17]: does this need to be a qwidget?

File: tools/clib/FrameReader.hpp [hunk 34]
```diff
@@ -30,15 +30,17 @@ class FrameReader {
   int getRGBSize() { return width*height*3; }
   void loaderThread();
   void cacherThread();
+
+  //TODO: get this from the actual frame
+	int width = 1164;
+	int height = 874;
```
Comment (adeebshihadeh) [inline tools/clib/FrameReader.hpp:36]: spaces

File: selfdrive/ui/ui.hpp [hunk 35]
```diff
@@ -165,6 +165,7 @@ typedef struct UIState {
   bool awake;
 
   bool sidebar_collapsed;
+  bool replay;
```
Comment (adeebshihadeh) [inline selfdrive/ui/ui.hpp:168]: remove this

File: selfdrive/ui/ui.cc [hunk 36]
```diff
@@ -349,7 +350,9 @@ static void update_status(UIState *s) {
 
 void ui_update(UIState *s) {
   update_params(s);
-  update_sockets(s);
+  if(!s->replay){
```
Comment (adeebshihadeh) [inline selfdrive/ui/ui.cc:353]: remove this

File: selfdrive/ui/ui.cc [hunk 37]
```diff
@@ -56,6 +56,7 @@ void ui_init(UIState *s) {
   });
 
   s->scene.started = false;
+  s->replay = false;
```
Comment (adeebshihadeh) [inline selfdrive/ui/ui.cc:59]: remove this

File: selfdrive/ui/replay/replay.cc [hunk 38]
```diff
@@ -0,0 +1,70 @@
+#include "replay.hpp"
+
+Replay::Replay(QString route_, int seek) : route(route_) {
+  unlogger = new Unlogger(&events, &events_lock, &frs, seek);
+  seg_add = 0;
+
+  http = new HttpRequest(this, "https://api.commadotai.com/v1/route/" + route.replace("/", "|") + "/files", "Route_HttpRequest");
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.cc:7]: it was entered with the pipe, and now you're just reverting something done in replay/main.cc

File: selfdrive/ui/replay/replay.cc [hunk 39]
```diff
@@ -0,0 +1,72 @@
+#include "replay.hpp"
+
+Replay::Replay(QString route_, int seek) : route(route_) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.cc:3]: why are there two different seeks? there's one in stream too

File: selfdrive/ui/replay/replay.cc [hunk 40]
```diff
@@ -0,0 +1,72 @@
...
+  unlogger = new Unlogger(&events, &events_lock, &frs, seek);
+  seg_add = 0;
+
+  http = new HttpRequest(this, "https://api.commadotai.com/v1/route/" + route + "/files", "Route_HttpRequest");
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.cc:7]: remove the cache key

**Pull request merged**